Rewrite and simplify the paser; add proper layout.

kosmikus opened the pull request:
[matches PR https://github.com/ucsd-progsys/liquidhaskell/pull/1704]

This is a major restructuring, even though more work remains to do.

The most important changes are:

- Use megaparsec instead of parsec.
- Change the handling of internal parser state.
- Improve error reporting somewhat.
- Improve location tracking somewhat.
- Add layout information to the parser state.
- Export layout block parser combinators.
- Add more coments and TODOs.

ranjitjhala commented on src/Language/Fixpoint/Parse.hs:
```diff
@@ -579,18 +960,19 @@ sortFunP
+   =  parens sortP -- parenthesised sort, starts with "("
+  <|> (reserved "func" >> funcSortP) -- function sort, starts with "func"
+  <|> (fAppTC listFTyCon . pure <$> brackets sortP)
+  <|> bvSortP -- Andres: this looks unreachable, as it starts with "("
```
hmm that's strange, but we do have some tests for it, search for `(BitVec Size32)` in tests/ so if they pass it should be ok?

kosmikus replied:
I think `bvSortP` is indeed unreachable. But if you look at how `bvSortP` is defined, then it is:
```
bvSortP :: Parser Sort
bvSortP = mkSort <$> (bvSizeP "Size32" S32 <|> bvSizeP "Size64" S64)
  where
    bvSizeP ss s = do
      parens (reserved "BitVec" >> reserved ss)
      return s
```
So it recognises `(BitVec Sort32)` and `(BitVec Sort64)`. The strings also parse via the `(fAppTC <$> fTyConP <*> appArgsP)` option of this parser. So I think this is just redundant.

kosmikus replied:
Actually, I think I'll comment out this branch, as I'm rather confident it is not having any effect anyway, and tests pass withou it.

ranjitjhala commented:
Hi @kosmikus -- this is great, shall we merge? Was there a matching PR for `liquidhaskell` too?

kosmikus committed:
Rewrite and simplify the paser; add proper layout.

This is a major restructuring, even though more work remains to do.

The most important changes are:

- Use megaparsec instead of parsec.
- Change the handling of internal parser state.
- Improve error reporting somewhat.
- Improve location tracking somewhat.
- Add layout information to the parser state.
- Export layout block parser combinators.
- Add more coments and TODOs.

kosmikus commented:
> Hi @kosmikus -- this is great, shall we merge? Was there a matching PR for `liquidhaskell` too?

@ranjitjhala I'm happy to merge this (and ucsd-progsys/liquidhaskell#1704) whenever you are ...
